book/src/concepts/system.md

```diff
@@ -290,6 +290,153 @@ write_storage.remove(entity);
+ * `GameplayState` - which we enter after the main menu, while the game is running.
+
+The following example shows how to keep track of which state we are currently in.
+This allows us to do a bit of conditional logic in our systems to determine what to do depending on
+which state is currently active. And manipulating the states by setting flags:
+
+```rust,no_run,noplaypen
+use amethyst::prelude::*;
+
+#[derive(Clone, Copy)]
+enum CurrentState {
+    MainMenu,
+    Gameplay,
+}
+
+impl Default for CurrentState {
+    fn default() -> Self {
+        CurrentState::Gameplay
+    }
+}
+
+#[derive(Default)]
+struct Game {
+    continue_game: bool,
+    quit_game: bool,
+    open_menu: bool,
+    current_state: CurrentState,
+}
+
+struct GameplayState;
+
+impl<'a,'b> SimpleState<'a,'b> for GameplayState {
+    fn update(&mut self, data: StateData<GameData>) {
+        // If the `Game` resource has been set up to go back to the menu, pop the state so that
+        // we go back.
+
+        {
+            let gameplay = data.world.write_resource::<Game>();
+
+            if gameplay.open_menu {
+                // Reset the flag to prevent immediately open the menu the next time it's opened.
+                gameplay.open_menu = false;
+                return Trans::Push(Box::new(MainMenuState));
+            }
+        }
+
+        Trans::None
+    }
+
+    fn on_resume(&mut self, data: StateData<GameData>) {
+        // mark that the current state is a gameplay state.
+        data.world.write_resource::<Game>().current = CurrentState::Gameplay;
+    }
+}
+
+struct MainMenuState;
+
+impl<'a,'b> SimpleState<'a,'b> for MainMenuState {
+    fn update(&mut self, data: StateData<GameData>) {
+        {
+            let game = data.world.write_resource::<Game>();
+
+            // Continuing the game simply means popping back to the gameplay state.
+            if game.continue_game {
+                // Reset the flag to prevent immediately closing the menu the next time it's
+                // opened.
+                game.continue_game = false;
+                return Trans::Pop;
+            }
+
+            if game.quit_game {
+                // Note: no need to clean up :)
+                return Trans::Quit;
+            }
+        }
+
+        Trans::None
+    }
+
+    fn on_resume(&mut self, data: StateData<GameData>) {
+        // mark that the current state is a main menu state.
+        data.world.write_resource::<Game>().current = CurrentState::MainMenu;
+    }
+}
+```
+
+Let's say we want the player to be able to press escape to enter the menu.
+We modify our input handler to map the `open_menu` action to `Esc`, and we write the following
+system:
+
+```rust,no_run,noplaypen
+use amethyst::{
+    prelude::*,
+    ecs::{System, prelude::*},
+    input::InputHandler,
+};
+
+struct MyFirstSystem;
+
+impl<'s> System<'s> for MyFirstSystem {
+    type SystemData = (
+        Read<'s, InputHandler<String, String>>,
+        Write<'s, Gameplay>,
```

Comment:
Also, what you can do is react to `InputEvent<String>` events in the state's handle method. Those are the events emitted by the InputHandler. The strings you get is the key used in your input configuration (actions).

Note: We need to add `InputEvent<T = String>` to the default `StateEvent` of amethyst for events to actually be forwarded to the State::handle_event method. I opened an issue for this yesterday as a reminder.

Reply:
This does cover state transitions caused immediately by keyboard inputs, but not by other things, like the player dying because of a collision.

Honestly though, it seemed a bit off to me to handle input events both in states and in systems. I'd much rather have a clear signal to a new user to do it one way or another. I'd love more input on this though.